### PR TITLE
[Blacklist] Redesign and rework the blacklist

### PR DESCRIPTION
**Pull request by Sindrake**
This pull request replicates re621's extended blacklist to the vanilla site.
This includes both a redesign of the blacklist UI, and adding more usable metatags.

### Sidebar 

The blacklist UI in the sidebar had been made more clear and easy to use.
Before, it was not immediately obvious to some users that the crossed-out filters were disabled.
Muting the colors where appropriate, and adding checkboxes that reflect the filter's state, should help with that.

Additionally, the number of posts next to the blacklist header updates dynamically when individual filters are enabled or disabled, and reflects the total number of posts that had been hidden by the blacklist.

<br clear="left"/>

### Inline
inline

The inline blacklist UI looks more or less the same as before.
Under the hood, it is identical to the sidebar UI, just with some style changes.

### Persistence

Previous blacklist implementation could only save its state between different pages as all or nothing: either all filters are disabled, or all are enabled.

Now, if an individual filter is disabled on one page, it will remain disabled on another.
"Enable All Filters" clears the saved filter states on all pages. Additionally, the script will check that the filters in the disabled list actually still exist on startup, to prevent entries from getting stuck in the list despite being removed from the blacklist.

#### Search page 

User disables some filters on the search page.
<br clear="left"/>

#### Post page

When the user navigates to a post page, the same filters will remain disabled.
<br clear="left"/>

### Metatags <a id="metatags"></a>
The table below lists the available metatags.
Blue checkmarks indicate previous functionality, green ones – tags that had been added in this PR.

"Equality" indicates an exact match: `score:100`.
"Comparison" is relative: `score:>100`.
"Range" supports a range of values: `score:50..100`.

|                  | Equality | Comparison | Range |
|       ---      |     :---:    |        :---:        |   :---:   |
| id             |     ☑️      |        ✅        |    ✅   |
| status       |     ☑️      |                    |             |
| rating       |     ☑️      |                    |             |
| type          |     ✅      |                    |             |
| width        |     ☑️      |        ✅        |    ✅   |
| height      |     ☑️      |         ✅        |    ✅   |
| filesize <sup>1</sup>    |    ✅      |         ✅        |    ✅   |
| score        |     ☑️      |        ☑️        |    ✅   |
| favcount   |     ✅      |        ✅        |    ✅   |
| fav:me <sup>2</sup>   |     ☑️      |                    |             |
| user          |     ☑️      |                    |             |
| userid       |     ☑️      |         ✅        |    ✅   |
| username |     ✅      |                    |             |
| pool         |     ✅      |                    |             |
| tagcount  |     ✅      |         ✅        |    ✅   |

<sup>1</sup> `filesize` accepts both numerical values, or ones in `kb` or `mb`
<sup>2</sup> `fav:me` can only refer to the user's own favorites, same as before.

### Comments
The blacklist now supports proper comments, both inline and standalone.
Anything after the `#` symbol will be ignored by the blacklist.

All of the following are valid:
```
horse male solo # will remove later
rainbow lizard #scaliespe`
# artists I hate:
##### break #####
```

### Click to Reveal
On the post page, clicking on the blacklist placeholder will reveal the post without disabling any of the filters.

### CODE REVIEW

File: app/javascript/src/javascripts/utility/filter_util.js [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+      return a < parseFloat(token.value);
+    case "<=":
+      return a <= parseFloat(token.value);
+    case ">":
+      return a > parseFloat(token.value);
+    case ">=":
+      return a >= parseFloat(token.value);
+    case "..": {
+      const parts = token.value.split("..");
+      if (parts.length !== 2) return false;
```
Comment (DonovanDMC) [inline app/javascript/src/javascripts/utility/filter_util.js]: Within the site only one side needs to be provided, this acts like `>=` / `<=`

`..20` = <=20
`20..` = >=20

https://github.com/e621ng/e621ng/blob/17b99821e3f31ea640a70d51a9631e8584fc11e2/app/logical/parse_value.rb#L40-L41
https://github.com/e621ng/e621ng/blob/17b99821e3f31ea640a70d51a9631e8584fc11e2/app/logical/parse_value.rb#L49-L50

File: app/javascript/src/styles/common/blacklists.scss [hunk 2]
```diff
@@ -1,86 +1,176 @@
+/*** QUick Edit Form ***/
```
Comment (DonovanDMC) [inline app/javascript/src/styles/common/blacklists.scss]:
```suggestion
/*** Quick Edit Form ***/
```

**Review by DonovanDMC — CHANGES_REQUESTED** [review]

File: app/javascript/src/javascripts/utility/filter_util.js (hunk 1 above)
Comment (Sindrake) [reply app/javascript/src/javascripts/utility/filter_util.js]: That is a silly format.

**Review by DonovanDMC — APPROVED** [review]

**Comment by Sindrake** [comment]
Spent a ton of time optimizing performance.  
Old blacklist processes 320 posts in ~101ms, new one does the same in ~47ms.
Considering all the new features that were added, I am satisfied.

**Comment by mm12** [comment]
might be the best addition in years

**Comment by Sindrake** [comment]
Might be the buggiest addition in years.

**Pull request merged**